### PR TITLE
feat: Add outbox for audit logs

### PR DESCRIPTION
**Pull request by ebbeknudsen**
## Description

- Add outbox module
- USe outbox to publish auditlog's in `AuditLogger` (instead of doing it synchronously)
- Update `UnitOfWork` implementation to get list of `DbContext`'s from IoC container
- Add/update tests

**d002 EDI outbox table**
|Id|RecordId|CreatedBy|CreatedAt|ModifiedBy|ModifiedAt|RowVersion|Type|Payload|ProcessingAt|PublishedAt|FailedAt|ErrorMessage|ErrorCount|
|---|---|---|---|---|---|---|---|---|---|---|---|---|---|
|41e13707-ee84-4b80-a9ea-23fc8cda549d|1|5790001330583|2024-09-06 05:25:11.0830725|Unknown|2024-09-06 05:25:20.2193079|0x0000000000022AD6|AuditLogOutboxMessageV1|{&quot;LogId&quot;:&quot;a715f818-1962-4bae-81bf-a05a6afad12c&quot;,&quot;UserId&quot;:&quot;00000000-0000-1111-0000-000000000000&quot;,&quot;ActorId&quot;:&quot;00000000-0000-0000-0000-000000000001&quot;,&quot;SystemId&quot;:&quot;688b2dca-7231-490f-a731-d7869d33fe5e&quot;,&quot;Permissions&quot;:&quot;delegation:view, calculations:manage, settlement-reports:manage, grid-areas:manage, actors:manage, users:manage, users:view, user-roles:manage, imbalance-prices:manage, esett-exchange:manage, actor-credentials:manage, actor-master-data:manage, delegation:manage, users:reactivate, balance-responsibility:view, calculations:view, imbalance-prices:view&quot;,&quot;OccuredOn&quot;:&quot;2024-09-06T05:25:10.0823155Z&quot;,&quot;Activity&quot;:&quot;ArchivedMessagesSearch&quot;,&quot;Origin&quot;:&quot;https://app-b2cwebapi-edi-d-we-002.azurewebsites.net/ArchivedMessageSearch&quot;,&quot;Payload&quot;:{&quot;CreatedDuringPeriod&quot;:null,&quot;MessageId&quot;:&quot;83a0e14eef9346a4850f9e83a77f2a7e&quot;,&quot;SenderNumber&quot;:null,&quot;ReceiverNumber&quot;:null,&quot;DocumentTypes&quot;:null,&quot;BusinessReasons&quot;:null,&quot;IncludeRelatedMessages&quot;:false},&quot;AffectedEntityType&quot;:&quot;ArchivedMessage&quot;,&quot;AffectedEntityKey&quot;:null}|2024-09-06 05:25:20.0577649|2024-09-06 05:25:20.2183171|NULL|NULL|0|

**d002 revision log table**
|Id|LogId|ReceivedAt|UserId|ActorId|SystemId|Activity|Origin|Payload|OccurredOn|AffectedEntityType|AffectedEntityKey|Permissions|
|---|---|---|---|---|---|---|---|---|---|---|---|---|
|58|a715f818-1962-4bae-81bf-a05a6afad12c|2024-09-06 05:25:20.1546233 +00:00|00000000-0000-1111-0000-000000000000|00000000-0000-0000-0000-000000000001|688b2dca-7231-490f-a731-d7869d33fe5e|ArchivedMessagesSearch|https://app-b2cwebapi-edi-d-we-002.azurewebsites.net/ArchivedMessageSearch|{&quot;CreatedDuringPeriod&quot;:null,&quot;MessageId&quot;:&quot;83a0e14eef9346a4850f9e83a77f2a7e&quot;,&quot;SenderNumber&quot;:null,&quot;ReceiverNumber&quot;:null,&quot;DocumentTypes&quot;:null,&quot;BusinessReasons&quot;:null,&quot;IncludeRelatedMessages&quot;:false}|2024-09-06 05:25:10.0000000 +00:00|ArchivedMessage||delegation:view, calculations:manage, settlement-reports:manage, grid-areas:manage, actors:manage, users:manage, users:view, user-roles:manage, imbalance-prices:manage, esett-exchange:manage, actor-credentials:manage, actor-master-data:manage, delegation:manage, users:reactivate, balance-responsibility:view, calculations:view, imbalance-prices:view|

## References

- https://app.zenhub.com/workspaces/mosaic-60a6105157304f00119be86e/issues/gh/energinet-datahub/team-mosaic/287
- [x] https://github.com/Energinet-DataHub/dh3-environments/actions/runs/10732687192

## Checklist
- [ ] Should the change be behind a feature flag?
- [x] Can the feature be meaningfully disabled or circumvented if there are issues (e.g., database-breaking changes)?
- [x] Has it been considered whether data is being delivered to the wrong actor?
- [x] Subsystem test executed (dev_002/dev_003)
- [x] Is there time to monitor state of the release to Production?
- [x] Reference to the task

### CODE REVIEW

File: source/B2BApi.AppTests/Fixtures/Database/EdiDatabaseManager.cs [hunk 1]
```diff
@@ -16,27 +16,31 @@
...
 using Microsoft.EntityFrameworkCore;
+using NodaTime;
 
 namespace Energinet.DataHub.EDI.B2BApi.AppTests.Fixtures.Database;
 
 public class EdiDatabaseManager : SqlServerDatabaseManager<DbContext>
 {
     public EdiDatabaseManager()
-        : base("Edi")
+        : base($"Edi_{DateTime.Now:yyyyMMddHHmm}_")
```
Comment (lasrinnil) [inline source/B2BApi.AppTests/Fixtures/Database/EdiDatabaseManager.cs:26]: Why is this needed?

File: source/B2BApi/Functions/OutboxPublisher.cs [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+namespace Energinet.DataHub.EDI.B2BApi.Functions;
+
+public class OutboxPublisher(IOutboxProcessor outboxProcessor)
+{
+    private readonly IOutboxProcessor _outboxProcessor = outboxProcessor;
+
+    [Function(nameof(OutboxPublisher))]
+    public Task PublishOutboxAsync(
+        [TimerTrigger("*/10 * * * * *")] TimerInfo timerTimerInfo,
```
Comment (lasrinnil) [inline source/B2BApi/Functions/OutboxPublisher.cs:26]: Perhaps we could update this trigger to run at an time when we know the load on the system is the lowest.

Or maybe just it should run once every hour or so - or is it because we fear that the log grows larger than we can process it?

File: source/B2BApi/HostFactory.cs [hunk 3]
```diff
@@ -88,7 +90,15 @@ public static IHost CreateHost(TokenValidationParameters tokenValidationParamete
                         .AddOutgoingMessagesModule(context.Configuration)
                         .AddProcessModule(context.Configuration)
                         .AddMasterDataModule(context.Configuration)
-                        .AddDataAccessUnitOfWorkModule(context.Configuration);
+                        .AddDataAccessUnitOfWorkModule()
+
+                        // Audit log (outbox publisher)
```
Comment (lasrinnil) [inline source/B2BApi/HostFactory.cs:95]:
```suggestion
                        // Audit log (outbox publisher)
```

This comment seems redundant to me.

File: source/B2BApi/HostFactory.cs [hunk 4]
```diff
@@ -88,7 +90,15 @@ public static IHost CreateHost(TokenValidationParameters tokenValidationParamete
...
+                        .AddAuditLogOutboxPublisher()
+
+                        // Outbox module and outbox processing
```
Comment (lasrinnil) [inline source/B2BApi/HostFactory.cs:98]:
```suggestion
                        // Outbox module and outbox processing
```

The naming on `AddOutboxModule` and `AddOutboxModule` is suffice for me 💯

File: source/B2CWebApi.AppTests/B2CWebApi.AppTests.csproj [hunk 5]
```diff
@@ -35,6 +35,9 @@
         <None Update="integrationtest.local.settings.json">
             <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
         </None>
+        <None Update="integrationtest.local.settings.json">
```
Comment (lasrinnil) [inline source/B2CWebApi.AppTests/B2CWebApi.AppTests.csproj:38]: required?

File: source/BuildingBlocks.Infrastructure/DataAccess/ResilientTransaction.cs [hunk 6]
```diff
@@ -27,37 +27,50 @@ namespace Energinet.DataHub.EDI.BuildingBlocks.Infrastructure.DataAccess;
...
+
+                await transaction.CommitAsync().ConfigureAwait(false);
             }
 
-            await transaction.CommitAsync().ConfigureAwait(false);
+            // The transaction is only removed from the "firstDbContext" after it is disposed (????) so we need
+            // to manually remove the disposed transaction from all db contexts, else an exception will be thrown
+            // if one of the dbContexts are used later.
+            foreach (var dbContext in dbContexts)
+                await dbContext.Database.UseTransactionAsync(null).ConfigureAwait(false);
```
Comment (lasrinnil) [inline source/BuildingBlocks.Infrastructure/DataAccess/ResilientTransaction.cs:73]: ☎️

File: source/DataAccess.UnitOfWork/UnitOfWork.cs [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+using Microsoft.EntityFrameworkCore;
+
+namespace Energinet.DataHub.EDI.DataAccess.UnitOfWork;
+
+public sealed class UnitOfWork : IUnitOfWork
+{
+    private readonly IReadOnlyCollection<DbContext> _dbContexts;
+    // private readonly ProcessContext _processContext;
+    // private readonly ActorMessageQueueContext _actorMessageQueueContext;
+    // private readonly IncomingMessagesContext _incomingMessagesContext;
```
Comment (lasrinnil) [inline source/DataAccess.UnitOfWork/UnitOfWork.cs:26]: remove

File: source/Outbox.Application/Extensions/DependencyInjection/OutboxExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+        return services;
+    }
+
+    /// <summary>
+    /// Add services required for processing and publishing outbox messages.
+    /// <remarks>Requires <see cref="AddOutboxModule"/> to be registered as well</remarks>
+    /// </summary>
+    public static IServiceCollection AddOutboxProcessor(
+        this IServiceCollection services)
+    {
```
Comment (lasrinnil) [inline source/Outbox.Application/Extensions/DependencyInjection/OutboxExtensions.cs:59]: Should we throw an exception if the dependencies from `AddOutboxModule` have not been registered?

**Review by lasrinnil — DISMISSED** [review]
Great work 🎆 🎆

File: source/AuditLog/AuditLogOutbox/AuditLogOutboxPublisher.cs [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    private readonly IAuditLogClient _auditLogClient = auditLogClient;
+    private readonly ISerializer _serializer = serializer;
+
+    public bool CanPublish(string type) => type.Equals(AuditLogOutboxMessageV1.OutboxMessageType, StringComparison.OrdinalIgnoreCase);
+
+    public async Task PublishAsync(string serializedPayload)
+    {
+        ArgumentException.ThrowIfNullOrEmpty(serializedPayload);
+
+        var payload = _serializer.Deserialize<AuditLogOutboxMessageV1Payload>(serializedPayload)
```
[the remark this answers is not in the document]
Comment (ebbeknudsen) [reply source/AuditLog/AuditLogOutbox/AuditLogOutboxPublisher.cs:33]: This test should already exist (`Given_AuditLogOutboxMessageV1PayloadJson_When_PublishAuditLogOutboxMessage_Then_MessageIsCorrectlyPublished`)

File: source/B2BApi.AppTests/Fixtures/Database/EdiDatabaseManager.cs (hunk 1 above)
Comment (ebbeknudsen) [reply source/B2BApi.AppTests/Fixtures/Database/EdiDatabaseManager.cs:26]: It isn't needed, it just makes it easier to find the correct/newest database used by the running test (if you have multiple databases because of interrupting a test before dispose() deletes the database again)

File: source/B2BApi/Functions/OutboxPublisher.cs (hunk 2 above)
Comment (ebbeknudsen) [reply source/B2BApi/Functions/OutboxPublisher.cs:26]: Do you think processing outbox messages will impact how well the system is running? Personally i just want them sent as quickly as possible, so the "asynchronous" action happens as close to synchronous action as possible, unless there is a good reason to delay it. Preferably they would be processed immediatly after being commited to the outbox table, but this introduces a delay of up to 10 seconds.

File: source/B2BApi/HostFactory.cs (hunk 3 above)
Comment (ebbeknudsen) [reply source/B2BApi/HostFactory.cs:95]: I can remove the comments, I just followed how the rest of the dependency injection setup is written.
image

File: source/Outbox.Application/Extensions/DependencyInjection/OutboxExtensions.cs (hunk 8 above)
Comment (ebbeknudsen) [reply source/Outbox.Application/Extensions/DependencyInjection/OutboxExtensions.cs:59]: ☎️

File: source/B2BApi/Functions/OutboxPublisher.cs (hunk 2 above)
Comment (lasrinnil) [reply source/B2BApi/Functions/OutboxPublisher.cs:26]: Agreed - OutboxMessages should be executed as soon as possible.

**Review by lasrinnil — APPROVED** [review]
